overlord/assertstate/assertstate.go

```diff
@@ -887,10 +891,97 @@ func TryEnforceValidationSets(st *state.State, validationSets []string, userID i
 	return addCurrentTrackingToValidationSetsHistory(st)
 }
 
-// EnforceValidationSet tries to fetch the given validation set and enforce it.
+// EnforceValidationSets takes a map from validation set specifiers in the
+// foo/bar=N syntax to ValidationSet assertions and enforces the sets, if possible.
+// It doesn't fetch any validation sets.
+func EnforceValidationSets(st *state.State, vsMap map[string]*asserts.ValidationSet, snaps []*snapasserts.InstalledSnap, ignoreValidation map[string]bool, userID int) error {
+	user, err := userFromUserID(st, userID)
+	if err != nil {
+		return err
+	}
+
+	sto := snapstate.Store(st, nil)
+	retrieve := func(ref *asserts.Ref) (asserts.Assertion, error) {
+		return sto.Assertion(ref.Type, ref.PrimaryKey, user)
+	}
+
+	db := cachedDB(st)
+	batch := asserts.NewBatch(nil)
+
+	extraVs := make([]*asserts.ValidationSet, 0, len(vsMap))
+	newTracking := make([]*ValidationSetTracking, 0, len(vsMap))
+	for vsStr, vs := range vsMap {
+		_, _, pinnedSeq, err := snapasserts.ParseValidationSet(vsStr)
+		if err != nil {
+			return err
+		}
+
+		if pinnedSeq != 0 && pinnedSeq != vs.Sequence() {
+			// the assertions we get here come from a previous attempt at enforcing which
+			// obtained them with the same valset string so this shouldn't be possible
+			return fmt.Errorf("internal error: trying to enforce validation set with sequence point different than pinned")
+		}
+
+		tr := &ValidationSetTracking{
+			AccountID: vs.AccountID(),
+			Name:      vs.Name(),
+			Mode:      Enforce,
+			Current:   vs.Sequence(),
+			// may be 0 meaning no pinning
+			PinnedAt: pinnedSeq,
+		}
+
+		newTracking = append(newTracking, tr)
+		extraVs = append(extraVs, vs)
+
+		for _, prereq := range vs.Prerequisites() {
+			fetching := func(f asserts.Fetcher) error {
+				if err := f.Fetch(prereq); err != nil {
+					return fmt.Errorf("cannot find prerequisite for validation set %s: %v", vsStr, err)
+				}
+
+				return nil
+			}
+
+			st.Unlock()
+			err = batch.Fetch(db, retrieve, fetching)
+			st.Lock()
+			if err != nil {
+				return err
+			}
+		}
+
+		batch.Add(vs)
```

Comment:
instead of going over the prerequisites explicitly this can be done by doing f.Save(vs) in the fetching function